java/languages.antlr/src/org/netbeans/modules/languages/antlr/AntlrParserResult.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.languages.antlr;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.antlr.v4.runtime.ANTLRErrorListener;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.Parser;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.tree.ParseTreeListener;
+import org.netbeans.modules.csl.api.Error;
+import org.netbeans.modules.csl.api.OffsetRange;
+import org.netbeans.modules.csl.api.Severity;
+import org.netbeans.modules.csl.spi.DefaultError;
+import org.netbeans.modules.csl.spi.ParserResult;
+import org.netbeans.modules.parsing.api.Snapshot;
+import org.netbeans.modules.parsing.api.Source;
+import org.openide.filesystems.FileObject;
+
+/**
+ *
+ * @author lkishalmi
+ */
+public abstract class AntlrParserResult<T extends Parser> extends ParserResult {
+
+    public final List<DefaultError> errors = new ArrayList<>();
+    public final Map<String, Reference> references = new TreeMap<>();
+
+    public final List<OffsetRange> folds = new ArrayList<>();
+    public final List<AntlrStructureItem> structure = new ArrayList<>();
+
+    final Deque<T> parsingQueue = new LinkedList<>();
+
+    final AtomicBoolean finished = new AtomicBoolean();
+
+    public AntlrParserResult(Snapshot snapshot) {
+        super(snapshot);
+        addParseTask(snapshot, false);
+    }
+    
+
+
+    public AntlrParserResult get() {
+        while (!parsingQueue.isEmpty()) {
```

Comment:
This looked hacky to me. If I understand this correctly, this is done to evaluate imported grammar parts. I only glanced at the antlr4 documentation, but it reads to me, that the files should be parsed independently. If I remember correctly the indexer infrastructure should offer the tools to lookup data from other parsed files, i.e. when code completion is invoked in a file, that imports another, completion needs to lookup the completion from the local file and merge that with data from the index.

Reply:
Yes, that's true, and then someone has to implement the indexing support. Maybe in another iteration. I could add a TODO here.

Reply:
I had another thought and it turned out, that I was right: There is the potential that the parser integration recurses indefinitely, if two grammars refer to themselves in a circular fashion.

Reply:
I'm looking into indexing implementation.

Reply:
Wow! Would be great. FYI. Just committed the Language split.

Reply:
Added a patch for the endless loop and a TODO comment.